### ISSUE TITLE
Add support for on-demand file checksum recalculation

### ISSUE THREAD

**Issue by eduardnegru** [xrootd/xrootd#1161]
According to #1160, there is no way to force a server to recompute a checksum once it has been computed. Thus, if one client queries the checksum (for example via xrdfs query checksum command), it will be extracted from the files's extended attributes and returned to the client.

Add support for on-demand checksum recalculation, along with a timestamp of the last recalculation. The timestamp can be returned as metadata information or it can be accessed through a xrdfs stat call.

Cheers.

**Comment by eduardnegru** [comment]
adding @costing

**Comment by Jo-stfc** [comment]
any updates on this? encountered similar issues as a result of disabling stale checksums server side (due to file modification times being inaccurate with Ceph) when requesting checksums mid transfer

**Comment by abh3** [comment]
I'm sorry to report that we have had no action on this. We could potentially prohibit querying the checksum if the file is open for writing. However, it's not clear how many workflow force a checksum prior to closing the file and such a change would disrupt them. Recording the file size would create havoc with all pre-existing checksum and we would need a complicated migration strategy. However, I am surprised that the modification time check did not force a recalculation unless the request was very close to the end of the copy. I suppose another question here is why is someone asking for a checksum while the file is in transit?

**Comment by Jo-stfc** [comment]
the modification time check was disabled at RAL (due to ceph updating the modification time for reads and stats). I'm not sure why they might have asked for a checksum while in transit as this case appears to be quite rare (1 occurrence in ~6 months of use)
but it could be due to request latency? being able to force a checksum recalculation would allow us to rectify those cases as they occur instead of declaring the file as lost

**Comment by abh3** [comment]
Yes, we added an option to checksum query to disable the modification test (that will appear in R6). Wed o not have a solution for queries before the file is complete. I'm not sure that a lot of effort is warranted for an issue that occurs once every 6 months. However, the easiest solution would be to prohibit checksum queries while the file is open for writing and do so by configuration so that we don't interfere  with pre-existing workflows. So, your call -- worth spending time on a twice a year issue?

**Comment by Jo-stfc** [comment]
ok. I'll see if I can address this next year along with a streamed checksum option, but will leave this for now

**Comment by amadio** [comment]
@abh3 I think just clearing the attribute (or recomputing and storing the new checksum) upon close for writing would solve the problem. What do you think?

**Comment by Jo-stfc** [comment]
thanks for the suggestion @amadio indeed that works on a patch I've made on the xrdceph close.

**Comment by abh3** [comment]
The general solution for XRootD is not as simple as it seems as we support multiple checksums and all of them would have to be deleted. I'm still figuring out what would be the best option here as there are several approaches and wholesale deletion may not be the best option. Plus, people to open a file for writing and never actually write to the file, so that's another nit in the scheme.
